Unstable test TestRepairTable

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

Can be found in our CI, this is a flaky test

### 2. What did you expect to see? (Required)

CI stable

### 3. What did you see instead (Required)

```
--- FAIL: TestRepairTable (1.05s)
    util.go:36:
                Error Trace:        /Users/tangliu/eng/tidb/pkg/testkit/external/util.go:36
                                                            /Users/tangliu/eng/tidb/pkg/ddl/repair_table_test.go:137
                Error:              Received unexpected error:
                                    [schema:1146]Table 'test.origin' doesn't exist
                Test:               TestRepairTable
FAIL
run all tasks takes 16m14.210871666s
panic: runtime error: index out of range [28] with length 25

goroutine 1 [running]:
[github.com/pingcap/failpoint/code.Restorer.Restore](http://github.com/pingcap/failpoint/code.Restorer.Restore)({{0x140000a0078?, 0x1400038a0d0?}})
        /Users/tangliu/eng/go/pkg/mod/[github.com/pingcap/failpoint@v0.0.0-20240527053858-9b3b6e34194a/code/restorer.go:114](http://github.com/pingcap/failpoint@v0.0.0-20240527053858-9b3b6e34194a/code/restorer.go:114) +0x49c
main.restoreFiles({0x140000b0020, 0x492, 0x140000d9f28?})
        /Users/tangliu/eng/go/pkg/mod/[github.com/pingcap/failpoint@v0.0.0-20240527053858-9b3b6e34194a/failpoint-ctl/main.go:94](http://github.com/pingcap/failpoint@v0.0.0-20240527053858-9b3b6e34194a/failpoint-ctl/main.go:94) +0x5c
main.main()
        /Users/tangliu/eng/go/pkg/mod/[github.com/pingcap/failpoint@v0.0.0-20240527053858-9b3b6e34194a/failpoint-ctl/main.go:82](http://github.com/pingcap/failpoint@v0.0.0-20240527053858-9b3b6e34194a/failpoint-ctl/main.go:82) +0x4e0
```


### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

master 

```
commit c9c865ced461b01b08095c760b705642c7f6c838 (HEAD -> master, origin/master, origin/HEAD)
Author: D3Hunter <jujj603@gmail.com>
Date:   Wed May 29 15:58:21 2024 +0800

    test: fix flaky test TestParallelDDL (#53636)
    
    close pingcap/tidb#39222
```

--- a/pkg/domain/domain.go
+++ b/pkg/domain/domain.go
@@ -283,8 +283,10 @@ func (do *Domain) loadInfoSchema(startTS uint64) (infoschema.InfoSchema, bool, i
 		is, relatedChanges, diffTypes, err := do.tryLoadSchemaDiffs(m, currentSchemaVersion, neededSchemaVersion, startTS)
 		if err == nil {
 			infoschema_metrics.LoadSchemaDurationLoadDiff.Observe(time.Since(startTime).Seconds())
+			isV2, _ := infoschema.IsV2(is)
 			do.infoCache.Insert(is, schemaTs)
 			logutil.BgLogger().Info("diff load InfoSchema success",
+				zap.Bool("isV2", isV2),
 				zap.Int64("currentSchemaVersion", currentSchemaVersion),
 				zap.Int64("neededSchemaVersion", neededSchemaVersion),
 				zap.Duration("start time", time.Since(startTime)),
@@ -436,7 +438,15 @@ func (*Domain) fetchSchemasWithTables(schemas []*model.DBInfo, m *meta.Meta, don
 			infoschema.ConvertCharsetCollateToLowerCaseIfNeed(tbl)
 			// Check whether the table is in repair mode.
 			if domainutil.RepairInfo.InRepairMode() && domainutil.RepairInfo.CheckAndFetchRepairedTable(di, tbl) {
-				continue
+				if tbl.State == model.StatePublic {
+					// If the state is public, it means that the DDL job is done, but the table
+					// haven't been deleted from the repair table list.
+					// Since the repairment is done and table is visible, we should load it.
+				} else {
+					// Do not load it because we are reparing the table and the table info could be `bad`
+					// before repair is done.
+					continue
+				}
 			}
 			di.Tables = append(di.Tables, tbl)
 		}

--- a/pkg/domain/domain.go
+++ b/pkg/domain/domain.go
@@ -438,7 +438,7 @@ func (*Domain) fetchSchemasWithTables(schemas []*model.DBInfo, m *meta.Meta, don
 			infoschema.ConvertCharsetCollateToLowerCaseIfNeed(tbl)
 			// Check whether the table is in repair mode.
 			if domainutil.RepairInfo.InRepairMode() && domainutil.RepairInfo.CheckAndFetchRepairedTable(di, tbl) {
-				if !tbl.State == model.StatePublic {
+				if tbl.State != model.StatePublic {
 					// Do not load it because we are reparing the table and the table info could be `bad`
 					// before repair is done.
 					continue

--- a/pkg/domain/domain.go
+++ b/pkg/domain/domain.go
@@ -438,15 +438,14 @@ func (*Domain) fetchSchemasWithTables(schemas []*model.DBInfo, m *meta.Meta, don
 			infoschema.ConvertCharsetCollateToLowerCaseIfNeed(tbl)
 			// Check whether the table is in repair mode.
 			if domainutil.RepairInfo.InRepairMode() && domainutil.RepairInfo.CheckAndFetchRepairedTable(di, tbl) {
-				if tbl.State == model.StatePublic {
-					// If the state is public, it means that the DDL job is done, but the table
-					// haven't been deleted from the repair table list.
-					// Since the repairment is done and table is visible, we should load it.
-				} else {
+				if !tbl.State == model.StatePublic {
 					// Do not load it because we are reparing the table and the table info could be `bad`
 					// before repair is done.
 					continue
 				}
+				// If the state is public, it means that the DDL job is done, but the table
+				// haven't been deleted from the repair table list.
+				// Since the repairment is done and table is visible, we should load it.
 			}
 			di.Tables = append(di.Tables, tbl)
 		}

